pkg/defkinds/nodejs/builder.go

```diff
@@ -0,0 +1,299 @@
+		return state, img, err
+	}
+
+	stageName := buildOpts.Stage
+	isWebserverBuild := false
+	if strings.HasPrefix(stageName, "webserver-") {
+		isWebserverBuild = true
+		stageName = strings.TrimPrefix(stageName, "webserver-")
+	}
+
+	stageDef, err := def.ResolveStageDefinition(stageName, true)
+	if err != nil {
+		return state, img, xerrors.Errorf("could not resolve stage %q: %w", stageName, err)
+	}
+
+	if isWebserverBuild && *stageDef.Dev {
+		return state, img, xerrors.Errorf("webserver cannot be built from dev stages")
+	}
+
+	state, img, err = h.buildNodeJS(ctx, def, stageDef, buildOpts)
+	if err != nil {
+		err = xerrors.Errorf("could not build nodejs stage: %w", err)
+		return state, img, err
+	}
+
+	if !isWebserverBuild {
+		return state, img, nil
+	}
+
+	state, img, err = h.buildWebserver(ctx, def, state, buildOpts)
+	if err != nil {
+		err = xerrors.Errorf("could not build webserver stage: %w", err)
+		return state, img, err
+	}
+
+	return state, img, nil
+}
+
+func (h *NodeJSHandler) buildNodeJS(
+	ctx context.Context,
+	def Definition,
+	stageDef StageDefinition,
+	buildOpts builddef.BuildOpts,
+) (llb.State, *image.Image, error) {
+	state := llbutils.ImageSource(def.Locks.BaseImage, true)
+	baseImg, err := image.LoadMeta(ctx, def.Locks.BaseImage)
+	if err != nil {
+		return state, nil, xerrors.Errorf("loading %q metadata: %w", def.Locks.BaseImage, err)
+	}
+
+	img := image.CloneMeta(baseImg)
+	img.Config.Labels[builddef.ZbuildLabel] = "true"
+
+	state, err = llbutils.InstallSystemPackages(state, llbutils.APT, stageDef.Locks.SystemPackages)
+	if err != nil {
+		return state, img, xerrors.Errorf("failed to add \"install system pacakges\" steps: %w", err)
+	}
+
+	state = llbutils.CopyExternalFiles(state, stageDef.ExternalFiles)
+	state = llbutils.Mkdir(state, "1000:1000", "/app")
+	state = state.User("1000")
+	state = state.Dir("/app")
+
+	state = h.globalPackagesInstall(stageDef, state, buildOpts)
+
+	if *stageDef.Dev == false {
+		state = h.yarnInstall(stageDef, state, buildOpts)
+		state = h.copySourceDirs(stageDef, state, buildOpts)
+		state = h.build(stageDef, state)
+	}
+
+	setImageMetadata(stageDef, state, img)
+
+	return state, img, nil
+}
+
+func setImageMetadata(stageDef StageDefinition, state llb.State, img *image.Image) {
+	for _, dir := range stageDef.StatefulDirs {
+		fullpath := dir
+		if !path.IsAbs(fullpath) {
+			fullpath = path.Join("/app", dir)
+		}
+
+		img.Config.Volumes[fullpath] = struct{}{}
+	}
+
+	if *stageDef.Healthcheck {
+		img.Config.Healthcheck = &image.HealthConfig{
+			Test:     []string{"CMD", "http_proxy= test \"$(curl --fail http://127.0.0.1/_ping)\" = \"pong\""},
+			Interval: 10 * time.Second,
+			Timeout:  1 * time.Second,
+			Retries:  3,
+		}
+	}
+
+	nodeEnv := "development"
+	if *stageDef.Dev == false {
+		nodeEnv = "production"
+	}
+
+	img.Config.User = "1000"
+	img.Config.WorkingDir = "/app"
+	img.Config.Env = []string{
+		"PATH=" + getEnv(state, "PATH"),
+		"NODE_ENV=" + nodeEnv,
```

Comment:
for prod env, we usually set the `GENERATE_SOURCEMAP` env var to `false`. However, I think it's very related to react-create-app .

Reply:
Do you know if it works as expected when the env var is declared in the `.env` file?

Reply:
:thinking: I'd say no as the  `.env` var is not sourced at build time (only at runtime).

Reply:
Not for react, `.env` files are always copied into the image at build-time as webpack/CRA are replacing `process.env...` by actual values :wink:

Reply:
I'll try to add `GENERATE_SOURCEMAP=false` to the `.env` used by api-platform demo client and see if it's working. Otherwise I might end up adding a `build_env_vars` parameter or something like that :wink: